jena-arq/src/main/java/org/apache/jena/sparql/core/mosaic/TransactionalDistributed.java

```diff
@@ -0,0 +1,267 @@
+	 */
+	public <T> Future<T> submit(final DatasetGraph datasetGraph, final Supplier<T> supplier) {
+		return getWorker(datasetGraph).submit(() -> {
+			begin(datasetGraph);
+			return supplier.get();
+		});
+	}
+	
+	/**
+	 * Begin a transaction on the given DatasetGraph.
+	 */
+	protected DatasetGraph begin(final DatasetGraph datasetGraph) {
+		if (!isInTransaction()) {
+			throw new JenaException("No parent transaction");
+		}
+		try {
+			if (!children.contains(datasetGraph)) {
+				if (datasetGraph.supportsTransactions()) {
+					datasetGraph.begin(readWrite);
+				}
+				children.add(datasetGraph);
+			}
+			return datasetGraph;
+		} catch (final Exception exception) {
+			throw new JenaException(exception);
+		}
+	}
+	
+	/*
+	 * Transactional
+	 */
+	
+	@Override
+	public void begin(final ReadWrite readWrite) {
+		if (this.readWrite != null) {
+			throw new JenaException("Already in a transaction " + this.readWrite);
+		}
+		this.readWrite = readWrite;
+	}
+	
+	@Override
+	public void commit() {
+		List<Exception> threw = new LinkedList<>();
+		children
+			.forEach(datasetGraph -> {
+				try {
+					getWorker(datasetGraph).execute(() -> {
+						if (datasetGraph.supportsTransactions()) {
+							datasetGraph.commit();
+						}
+					});
+				} catch (final Exception exception) {
+					threw.add(exception);
+				}
+			});
+		this.children.clear();
+		this.readWrite = null;
+		if (!threw.isEmpty()) {
+			final JenaException jenaException = new JenaException();
+			threw.forEach((exception) -> {jenaException.addSuppressed(exception);});
+			throw jenaException;
+		}
+	}
+	
+	@Override
+	public void abort() {
+		List<Exception> threw = new LinkedList<>();
+		children
+			.forEach(datasetGraph -> {
+				try {
+					getWorker(datasetGraph).execute(() -> {
+						if (datasetGraph.supportsTransactions() && datasetGraph.supportsTransactionAbort()) {
+							datasetGraph.abort();
+						} else {
+							LOGGER.warn("Attempt to call Transactional.abort().", datasetGraph.getClass());
+						}
+					});
+				} catch (final Exception exception) {
+					threw.add(exception);
+				}
+			});
+		this.children.clear();
+		this.readWrite = null;
+		if (!threw.isEmpty()) {
+			final JenaException jenaException = new JenaException();
+			threw.forEach((exception) -> {jenaException.addSuppressed(exception);});
+			throw jenaException;
+		}
+	}
+	
+	@Override
+	public void end() {
+		if (readWrite != null && readWrite.equals(ReadWrite.WRITE)) {
+			LOGGER.warn("End without commit/abort");
+		}
+		List<Exception> threw = new LinkedList<>();
+		children
+			.forEach(datasetGraph -> {
+				try {
+					getWorker(datasetGraph).execute(() -> {
+						if (datasetGraph.supportsTransactions()) {
+							datasetGraph.end();
+						}
+					});
+				} catch (final Exception exception) {
+					threw.add(exception);
+				}
+			});
+		this.children.clear();
+		this.workers.values().forEach(threadDelegate -> {
```

Comment:
Just a style thing, but this could be point-free `this.workers.values().forEach(ThreadProxy::close)`.